File: skills/trip-planner/scripts/weather_helper.py
#!/usr/bin/env python3
"""
Weather helper for trip planning.
Fetches weather forecasts using wttr.in (free, no API key required).
"""

import argparse
import json
import sys
import urllib.request
import urllib.error
from datetime import datetime, timedelta


def fetch_weather(location: str, days: int = 3) -> dict:
    """
    Fetch weather forecast for a location.

    Args:
        location: City name or location (e.g., "New York", "London", "Tokyo")
        days: Number of days to forecast (max 3 for wttr.in free tier)

    Returns:
        Dictionary with weather data
    """
    # wttr.in JSON format
    # Format codes: https://github.com/chubin/wttr.in
    encoded_location = urllib.parse.quote(location)
    url = f"https://wttr.in/{encoded_location}?format=j1"

    try:
        req = urllib.request.Request(url, headers={'User-Agent': 'curl/7.64.1'})
        with urllib.request.urlopen(req, timeout=30) as response:
            data = json.loads(response.read().decode('utf-8'))
            return data
    except urllib.error.URLError as e:
        print(f"Error fetching weather: {e}", file=sys.stderr)
        sys.exit(1)
    except json.JSONDecodeError as e:
        print(f"Error parsing weather data: {e}", file=sys.stderr)
        sys.exit(1)


def get_forecast_summary(location: str, start_date: str = None, end_date: str = None) -> dict:
    """
    Get a summarized forecast for trip planning.

    Args:
        location: Destination city
        start_date: Trip start date (YYYY-MM-DD), defaults to today
        end_date: Trip end date (YYYY-MM-DD), defaults to start_date + 3 days

    Returns:
        Dictionary with summary suitable for packing decisions
    """
    data = fetch_weather(location)

    if not data or 'weather' not in data:
        return {"error": "Could not fetch weather data"}

    # Parse dates
    if start_date:
        start = datetime.strptime(start_date, "%Y-%m-%d")
    else:
        start = datetime.now()

    if end_date:
        end = datetime.strptime(end_date, "%Y-%m-%d")
    else:
        end = start + timedelta(days=3)

    # Current conditions
    current = data.get('current_condition', [{}])[0]

    # Analyze forecast days
    forecasts = data.get('weather', [])

    temps_f = []
    temps_c = []
    conditions = []
    rain_chances = []
    snow_chances = []

    for day in forecasts:
        temps_f.append(int(day.get('maxtempF', 70)))
        temps_f.append(int(day.get('mintempF', 50)))
        temps_c.append(int(day.get('maxtempC', 20)))
        temps_c.append(int(day.get('mintempC', 10)))

        # Check hourly for rain/snow
        for hour in day.get('hourly', []):
            rain_chances.append(int(hour.get('chanceofrain', 0)))
            snow_chances.append(int(hour.get('chanceofsnow', 0)))
            conditions.append(hour.get('weatherDesc', [{}])[0].get('value', ''))

    # Determine packing recommendations
    min_temp_f = min(temps_f) if temps_f else 50
    max_temp_f = max(temps_f) if temps_f else 70
    min_temp_c = min(temps_c) if temps_c else 10
    max_temp_c = max(temps_c) if temps_c else 20
    max_rain = max(rain_chances) if rain_chances else 0
    max_snow = max(snow_chances) if snow_chances else 0

    # Weather category for packing
    if min_temp_f <= 32:
        temp_category = "freezing"
    elif min_temp_f <= 45:
        temp_category = "cold"
    elif min_temp_f <= 60:
        temp_category = "cool"
    elif max_temp_f <= 75:
        temp_category = "mild"
    elif max_temp_f <= 85:
        temp_category = "warm"
    else:
        temp_category = "hot"

    return {
        "location": location,
        "forecast_generated": datetime.now().isoformat(),
        "temperature": {
            "min_f": min_temp_f,
            "max_f": max_temp_f,
            "min_c": min_temp_c,
            "max_c": max_temp_c,
            "category": temp_category
        },
        "precipitation": {
            "rain_chance_max": max_rain,
            "snow_chance_max": max_snow,
            "bring_umbrella": max_rain > 30,
            "bring_snow_gear": max_snow > 30 or min_temp_f <= 32
        },
        "packing_recommendations": {
            "winter_coat": temp_category in ["freezing", "cold"],
            "light_jacket": temp_category in ["cool", "mild"],
            "warm_layers": temp_category in ["freezing", "cold", "cool"],
            "shorts": temp_category in ["warm", "hot"],
            "umbrella": max_rain > 30,
            "waterproof_shoes": max_rain > 50 or max_snow > 30,
            "sunscreen": temp_category in ["warm", "hot", "mild"],
            "hat_warm": temp_category in ["freezing", "cold"],
            "hat_sun": temp_category in ["warm", "hot"],
            "gloves": temp_category in ["freezing", "cold"],
            "scarf": temp_category in ["freezing", "cold"]
        },
        "daily_forecasts": [
            {
                "date": day.get('date'),
                "high_f": day.get('maxtempF'),
                "low_f": day.get('mintempF'),
                "high_c": day.get('maxtempC'),
                "low_c": day.get('mintempC'),
                "description": day.get('hourly', [{}])[4].get('weatherDesc', [{}])[0].get('value', 'Unknown') if day.get('hourly') else 'Unknown'
            }
            for day in forecasts[:min(7, len(forecasts))]
        ],
        "summary": f"{temp_category.title()} weather expected. Temps: {min_temp_f}°F - {max_temp_f}°F ({min_temp_c}°C - {max_temp_c}°C). " +
                   (f"Rain likely ({max_rain}% chance). " if max_rain > 30 else "") +
                   (f"Snow possible ({max_snow}% chance). " if max_snow > 30 else "")
    }


def main():
    parser = argparse.ArgumentParser(description='Fetch weather forecast for trip planning')
    parser.add_argument('location', help='Destination city (e.g., "New York", "London")')
    parser.add_argument('--start', help='Trip start date (YYYY-MM-DD)')
    parser.add_argument('--end', help='Trip end date (YYYY-MM-DD)')
    parser.add_argument('--json', action='store_true', help='Output as JSON')

    args = parser.parse_args()

    forecast = get_forecast_summary(args.location, args.start, args.end)

    if args.json:
        print(json.dumps(forecast, indent=2))
    else:
        # Human-readable output
        print(f"\n🌤️  Weather Forecast for {forecast['location']}")
        print("=" * 50)
        print(f"\n{forecast['summary']}")
        print(f"\nTemperature Range: {forecast['temperature']['min_f']}°F - {forecast['temperature']['max_f']}°F")
        print(f"                   ({forecast['temperature']['min_c']}°C - {forecast['temperature']['max_c']}°C)")

        print("\n📅 Daily Forecast:")
        for day in forecast['daily_forecasts']:
            print(f"  {day['date']}: {day['low_f']}°F - {day['high_f']}°F | {day['description']}")

        print("\n🎒 Packing Recommendations:")
        recs = forecast['packing_recommendations']
        if recs['winter_coat']:
            print("  ✓ Winter coat / heavy jacket")
        if recs['light_jacket']:
            print("  ✓ Light jacket")
        if recs['warm_layers']:
            print("  ✓ Warm layers (sweaters, long sleeves)")
        if recs['shorts']:
            print("  ✓ Shorts / light clothing")
        if recs['umbrella']:
            print("  ✓ Umbrella / rain gear")
        if recs['waterproof_shoes']:
            print("  ✓ Waterproof shoes")
        if recs['gloves']:
            print("  ✓ Gloves")
        if recs['scarf']:
            print("  ✓ Scarf")
        if recs['hat_warm']:
            print("  ✓ Warm hat / beanie")
        if recs['hat_sun']:
            print("  ✓ Sun hat / cap")
        if recs['sunscreen']:
            print("  ✓ Sunscreen")
        print()


if __name__ == '__main__':
    import urllib.parse
    main()

File: skills/trip-planner/scripts/weather-helper.ts
#!/usr/bin/env bun
/**
 * Weather helper for trip planning.
 * Fetches weather forecasts using wttr.in (free, no API key required).
 *
 * Usage:
 *   bun scripts/weather-helper.ts "New York"
 *   bun scripts/weather-helper.ts "Tokyo" --json
 *   bun scripts/weather-helper.ts "London" --start 2026-01-15 --end 2026-01-20
 */

type TempCategory = "freezing" | "cold" | "cool" | "mild" | "warm" | "hot";

interface WttrHourly {
  chanceofrain?: string;
  chanceofsnow?: string;
  weatherDesc?: Array<{ value?: string }>;
}

interface WttrDay {
  date?: string;
  maxtempF?: string;
  mintempF?: string;
  maxtempC?: string;
  mintempC?: string;
  hourly?: WttrHourly[];
}

interface WttrResponse {
  current_condition?: Array<Record<string, unknown>>;
  weather?: WttrDay[];
}

interface ForecastSummary {
  location: string;
  forecast_generated: string;
  temperature: {
    min_f: number;
    max_f: number;
    min_c: number;
    max_c: number;
    category: TempCategory;
  };
  precipitation: {
    rain_chance_max: number;
    snow_chance_max: number;
    bring_umbrella: boolean;
    bring_snow_gear: boolean;
  };
  packing_recommendations: {
    winter_coat: boolean;
    light_jacket: boolean;
    warm_layers: boolean;
    shorts: boolean;
    umbrella: boolean;
    waterproof_shoes: boolean;
    sunscreen: boolean;
    hat_warm: boolean;
    hat_sun: boolean;
    gloves: boolean;
    scarf: boolean;
  };
  daily_forecasts: Array<{
    date: string | null;
    high_f: string | null;
    low_f: string | null;
    high_c: string | null;
    low_c: string | null;
    description: string;
  }>;
  summary: string;
}

async function fetchWeather(location: string): Promise<WttrResponse> {
  const url = `https://wttr.in/${encodeURIComponent(location)}?format=j1`;
  let res: Response;
  try {
    res = await fetch(url, {
      headers: { "User-Agent": "curl/7.64.1" },
      signal: AbortSignal.timeout(30_000),
    });
  } catch (err) {
    console.error(`Error fetching weather: ${(err as Error).message}`);
    process.exit(1);
  }
  if (!res.ok) {
    console.error(`Error fetching weather: HTTP ${res.status}`);
    process.exit(1);
  }
  try {
    return (await res.json()) as WttrResponse;
  } catch (err) {
    console.error(`Error parsing weather data: ${(err as Error).message}`);
    process.exit(1);
  }
}

function toInt(value: string | undefined, fallback: number): number {
  if (value === undefined) return fallback;
  const n = parseInt(value, 10);
  return Number.isNaN(n) ? fallback : n;
}

function describeDay(day: WttrDay): string {
  const hours = day.hourly;
  if (!hours || hours.length === 0) return "Unknown";
  // wttr.in returns 8 hourly slots (every 3 hours); index 4 is noon-ish.
  // Fall back to the middle of whatever we got if the day is short.
  const idx = hours.length > 4 ? 4 : Math.floor(hours.length / 2);
  return hours[idx]?.weatherDesc?.[0]?.value ?? "Unknown";
}

export function getForecastSummary(
  location: string,
  data: WttrResponse,
): ForecastSummary | { error: string } {
  if (!data || !data.weather) {
    return { error: "Could not fetch weather data" };
  }

  const forecasts = data.weather;
  const tempsF: number[] = [];
  const tempsC: number[] = [];
  const rainChances: number[] = [];
  const snowChances: number[] = [];

  for (const day of forecasts) {
    tempsF.push(toInt(day.maxtempF, 70));
    tempsF.push(toInt(day.mintempF, 50));
    tempsC.push(toInt(day.maxtempC, 20));
    tempsC.push(toInt(day.mintempC, 10));

    for (const hour of day.hourly ?? []) {
      rainChances.push(toInt(hour.chanceofrain, 0));
      snowChances.push(toInt(hour.chanceofsnow, 0));
    }
  }

  const minTempF = tempsF.length ? Math.min(...tempsF) : 50;
  const maxTempF = tempsF.length ? Math.max(...tempsF) : 70;
  const minTempC = tempsC.length ? Math.min(...tempsC) : 10;
  const maxTempC = tempsC.length ? Math.max(...tempsC) : 20;
  const maxRain = rainChances.length ? Math.max(...rainChances) : 0;
  const maxSnow = snowChances.length ? Math.max(...snowChances) : 0;

  const tempCategory: TempCategory =
    minTempF <= 32
      ? "freezing"
      : minTempF <= 45
        ? "cold"
        : minTempF <= 60
          ? "cool"
          : maxTempF <= 75
            ? "mild"
            : maxTempF <= 85
              ? "warm"
              : "hot";

  const isCold = tempCategory === "freezing" || tempCategory === "cold";
  const isCool = tempCategory === "cool";
  const isMild = tempCategory === "mild";
  const isWarm = tempCategory === "warm" || tempCategory === "hot";

  const summaryBits = [
    `${tempCategory[0].toUpperCase()}${tempCategory.slice(1)} weather expected.`,
    `Temps: ${minTempF}°F - ${maxTempF}°F (${minTempC}°C - ${maxTempC}°C).`,
  ];
  if (maxRain > 30) summaryBits.push(`Rain likely (${maxRain}% chance).`);
  if (maxSnow > 30) summaryBits.push(`Snow possible (${maxSnow}% chance).`);

  return {
    location,
    forecast_generated: new Date().toISOString(),
    temperature: {
      min_f: minTempF,
      max_f: maxTempF,
      min_c: minTempC,
      max_c: maxTempC,
      category: tempCategory,
    },
    precipitation: {
      rain_chance_max: maxRain,
      snow_chance_max: maxSnow,
      bring_umbrella: maxRain > 30,
      bring_snow_gear: maxSnow > 30 || minTempF <= 32,
    },
    packing_recommendations: {
      winter_coat: isCold,
      light_jacket: isCool || isMild,
      warm_layers: isCold || isCool,
      shorts: isWarm,
      umbrella: maxRain > 30,
      waterproof_shoes: maxRain > 50 || maxSnow > 30,
      sunscreen: isWarm || isMild,
      hat_warm: isCold,
      hat_sun: isWarm,
      gloves: isCold,
      scarf: isCold,
    },
    daily_forecasts: forecasts.slice(0, Math.min(7, forecasts.length)).map((day) => ({
      date: day.date ?? null,
      high_f: day.maxtempF ?? null,
      low_f: day.mintempF ?? null,
      high_c: day.maxtempC ?? null,
      low_c: day.mintempC ?? null,
      description: describeDay(day),
    })),
    summary: summaryBits.join(" ") + " ",
  };
}

interface CliArgs {
  location: string;
  start?: string;
  end?: string;
  json: boolean;
}

function parseArgs(argv: string[]): CliArgs {
  const args: CliArgs = { location: "", json: false };
  const positional: string[] = [];
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    if (a === "--json") {
      args.json = true;
    } else if (a === "--start") {
      args.start = argv[++i];
    } else if (a === "--end") {
      args.end = argv[++i];
    } else if (a === "-h" || a === "--help") {
      printHelp();
      process.exit(0);
    } else {
      positional.push(a);
    }
  }
  if (positional.length === 0) {
    printHelp();
    process.exit(2);
  }
  args.location = positional.join(" ");
  return args;
}

function printHelp(): void {
  console.log(
    `Usage: bun weather-helper.ts <location> [--start YYYY-MM-DD] [--end YYYY-MM-DD] [--json]\n` +
      `\n` +
      `  <location>   Destination city (quote multi-word names)\n` +
      `  --start      Trip start date (informational; wttr.in always returns the next 3 days)\n` +
      `  --end        Trip end date (informational)\n` +
      `  --json       Emit machine-readable JSON instead of the human summary\n`,
  );
}

function printHuman(forecast: ForecastSummary): void {
  console.log(`\n🌤️  Weather Forecast for ${forecast.location}`);
  console.log("=".repeat(50));
  console.log(`\n${forecast.summary}`);
  console.log(
    `\nTemperature Range: ${forecast.temperature.min_f}°F - ${forecast.temperature.max_f}°F`,
  );
  console.log(
    `                   (${forecast.temperature.min_c}°C - ${forecast.temperature.max_c}°C)`,
  );

  console.log("\n📅 Daily Forecast:");
  for (const day of forecast.daily_forecasts) {
    console.log(`  ${day.date}: ${day.low_f}°F - ${day.high_f}°F | ${day.description}`);
  }

  console.log("\n🎒 Packing Recommendations:");
  const recs = forecast.packing_recommendations;
  const labels: Array<[keyof typeof recs, string]> = [
    ["winter_coat", "Winter coat / heavy jacket"],
    ["light_jacket", "Light jacket"],
    ["warm_layers", "Warm layers (sweaters, long sleeves)"],
    ["shorts", "Shorts / light clothing"],
    ["umbrella", "Umbrella / rain gear"],
    ["waterproof_shoes", "Waterproof shoes"],
    ["gloves", "Gloves"],
    ["scarf", "Scarf"],
    ["hat_warm", "Warm hat / beanie"],
    ["hat_sun", "Sun hat / cap"],
    ["sunscreen", "Sunscreen"],
  ];
  for (const [key, label] of labels) {
    if (recs[key]) console.log(`  ✓ ${label}`);
  }
  console.log();
}

async function main(): Promise<void> {
  const args = parseArgs(process.argv.slice(2));
  const data = await fetchWeather(args.location);
  const result = getForecastSummary(args.location, data);

  if ("error" in result) {
    console.error(result.error);
    process.exit(1);
  }

  if (args.json) {
    console.log(JSON.stringify(result, null, 2));
  } else {
    printHuman(result);
  }
}

if (import.meta.main) {
  await main();
}
